File: apps/web/src/types/replays.ts
export type ReplayStatus =
  | "CREATED"
  | "READY"
  | "RUNNING"
  | "PAUSED"
  | "COMPLETED"
  | "STOPPED"
  | "FAILED";

export type ReplaySpeedMode = "MANUAL" | "X1" | "X10" | "X100";

export interface ReplayRun {
  id: string;
  idempotency_key: string;
  status: ReplayStatus;
  row_version: number;
  strategy_key: string;
  strategy_version: string;
  account_id: string | null;
  strategy_run_id: string | null;
  current_session_date: string | null;
  current_session_index: number;
  total_sessions: number;
  speed_mode: ReplaySpeedMode;
  bars_processed: number;
  signals_generated: number;
  orders_created: number;
  fills_generated: number;
  started_at: string | null;
  paused_at: string | null;
  completed_at: string | null;
  stopped_at: string | null;
  failed_at: string | null;
  error_code: string | null;
  error_message: string | null;
  last_heartbeat_at: string | null;
  worker_online: boolean;
  created_at: string;
  updated_at: string;
  replayed?: boolean;
  configuration?: ReplayConfiguration;
  final_summary?: Record<string, unknown>;
  integrity_summary?: Record<string, unknown>;
}

export interface ReplayConfiguration {
  speed_mode: ReplaySpeedMode;
  schema_version: number;
  execution: {
    parameters: Record<string, string | number | boolean>;
    instrument_ids: string[];
    start_at: string;
    end_at: string;
    initial_cash: string;
    order_type: string;
    time_in_force: string;
    fee_configuration: Record<string, string | number>;
    slippage_configuration: Record<string, string | number | null>;
    maximum_volume_participation: string | null;
  };
}

export interface ReplayPage {
  items: ReplayRun[];
  page: number;
  page_size: number;
  total: number;
}

export interface ReplayEvent {
  id: string;
  replay_run_id: string;
  sequence_number: number;
  event_type: string;
  business_time: string;
  occurred_at: string;
  instrument_id: string | null;
  related_entity_type: string | null;
  related_entity_id: string | null;
  summary: string;
  payload: Record<string, unknown>;
}

export interface ReplayEquityPoint {
  id: string;
  run_id: string;
  timestamp: string;
  cash: string;
  market_value: string;
  total_equity: string;
  gross_exposure: string;
  net_exposure: string;
  daily_return: string | null;
  cumulative_return: string;
  drawdown: string;
  positions_count: number;
  warnings: string[];
}

export interface ReplayState {
  status: ReplayStatus;
  row_version: number;
  cash: Record<string, unknown> | null;
  positions: Record<string, unknown>[];
  latest_equity: ReplayEquityPoint | null;
  [key: string]: unknown;
}

export interface ReplayIntegrity {
  replay_run_id: string;
  ok: boolean;
  checked_at: string;
  issues: string[];
  facts: Record<string, unknown>;
}

export interface CreateReplayRequest {
  strategy_key: string;
  parameters: Record<string, string | number | boolean>;
  instrument_ids: string[];
  start_at: string;
  end_at: string;
  initial_cash: string;
  order_type: string;
  time_in_force: string;
  fee_configuration: Record<string, string>;
  slippage_configuration: Record<string, string | null>;
  maximum_volume_participation: string | null;
  risk_configuration_reference: string;
  speed_mode: ReplaySpeedMode;
  data_source_code: string | null;
  idempotency_key: string;
}

File: apps/api/src/alphadesk_api/schemas/replays.py
"""RT01 public HTTP contracts."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from alphadesk_api.schemas.backtests import BacktestFeeBody, BacktestSlippageBody
from alphadesk_domain.market_reference import PriceAdjustmentMode


class ReplayCreateBody(BaseModel):
    strategy_key: str = Field(min_length=2, max_length=64)
    parameters: dict[str, str | int | bool] = Field(default_factory=dict)
    instrument_ids: list[UUID] = Field(min_length=1, max_length=200)
    start_at: datetime
    end_at: datetime
    initial_cash: str
    order_type: str = "LIMIT"
    time_in_force: str = "DAY"
    fee_configuration: BacktestFeeBody = Field(default_factory=BacktestFeeBody)
    slippage_configuration: BacktestSlippageBody = Field(default_factory=BacktestSlippageBody)
    maximum_volume_participation: str | None = None
    risk_configuration_reference: str = Field(default="r01-default-v1", max_length=128)
    speed_mode: str = "MANUAL"
    data_source_code: str | None = Field(default=None, min_length=1, max_length=64)
    idempotency_key: str = Field(min_length=1, max_length=128)
    strategy_price_adjustment_mode: PriceAdjustmentMode = PriceAdjustmentMode.RAW


class ReplayControlBody(BaseModel):
    idempotency_key: str = Field(min_length=1, max_length=128)
    expected_run_version: int = Field(ge=0)


class ReplaySpeedBody(ReplayControlBody):
    speed_mode: str


class ReplayRunResponse(BaseModel):
    id: UUID
    idempotency_key: str
    status: str
    row_version: int
    strategy_key: str
    strategy_version: str
    account_id: UUID | None
    strategy_run_id: UUID | None
    current_session_date: datetime | str | None
    current_session_index: int
    total_sessions: int
    speed_mode: str
    bars_processed: int
    signals_generated: int
    orders_created: int
    fills_generated: int
    started_at: datetime | None
    paused_at: datetime | None
    completed_at: datetime | None
    stopped_at: datetime | None
    failed_at: datetime | None
    error_code: str | None
    error_message: str | None
    last_heartbeat_at: datetime | None
    worker_online: bool
    created_at: datetime
    updated_at: datetime
    replayed: bool = False
    configuration: dict[str, Any] | None = None
    final_summary: dict[str, Any] | None = None
    integrity_summary: dict[str, Any] | None = None


class ReplayPageResponse(BaseModel):
    items: list[ReplayRunResponse]
    page: int
    page_size: int
    total: int


class ReplayCollectionResponse(BaseModel):
    items: list[dict[str, Any]]


class ReplayEventPageResponse(ReplayCollectionResponse):
    page: int
    page_size: int
    total: int


class ReplayStateResponse(BaseModel):
    state: dict[str, Any]


class ReplayIntegrityResponse(BaseModel):
    replay_run_id: UUID
    ok: bool
    checked_at: datetime
    issues: list[str]
    facts: dict[str, Any]
